lsp/nls/src/requests/completion.rs

```diff
@@ -29,18 +29,129 @@ use crate::{
 // We follow the path by traversing a term, type or contract which represents a record
 // and stop when there is nothing else on the path
 
+// Type or contract
+pub enum TorC {
+    T(Types),
+    C(RichTerm),
+}
+
+// #[derive(Eq, PartialEq, Debug, Ord, PartialOrd, Default)]
+struct IdentWithMeta {
+    ident: Ident,
+    info: TorC,
+    item: Option<LinearizationItem<Types>>,
+}
+
+impl From<Ident> for IdentWithMeta {
+    fn from(ident: Ident) -> Self {
+        IdentWithMeta {
+            ident,
+            info: TorC::T(Types(TypeF::Dyn)),
+            item: None,
+        }
+    }
+}
+
+impl From<&str> for IdentWithMeta {
+    fn from(ident: &str) -> Self {
+        IdentWithMeta {
+            ident: Ident::from(ident),
+            info: TorC::T(Types(TypeF::Dyn)),
+            item: None,
+        }
+    }
+}
+
+impl IdentWithMeta {
+    fn compute_detail(&self) -> String {
+        let info = || match &self.info {
+            TorC::T(ty) => ty.to_string(),
+            TorC::C(term) => term.to_string(),
+        };
+        let Some(item) = &self.item else {
+            return info()
+        };
+        item.meta
+            .as_ref()
+            .map(|meta| {
+                meta.types
+                    .as_ref()
+                    .map(|ty| ty.types.to_string())
+                    .or_else(|| {
+                        let result = meta
+                            .contracts
+                            .iter()
+                            .map(|contract| format!("{}", contract.label.types,))
+                            .collect::<Vec<_>>()
+                            .join(",");
+                        if result.is_empty() {
+                            None
+                        } else {
+                            Some(result)
+                        }
+                    })
+                    .unwrap_or_else(info)
+            })
+            .unwrap_or_else(info)
+    }
+
+    fn compute_completion_item_kind(&self) -> CompletionItemKind {
+        match &self.info {
+            TorC::T(ty) if ty.is_function_type() => CompletionItemKind::Function,
+            TorC::T(_) => CompletionItemKind::Property,
+            TorC::C(term) => match term.as_ref() {
+                Term::Fun(..) => CompletionItemKind::Function,
+                _ => CompletionItemKind::Property,
```

Comment:
I'm not sure about this one. Does it mean that if a term has a contract attached, and the contract is a function, then the term is considered to be documented as a function?

A contract is actually most of the time a function, but this doesn't tell anything about the value to which the contract is applied (and most of the time this function is hidden behind a variable, so this match case will never be taken anyway). I think there's not meaningful, simple, computable property able to check "is this contract checking for a function type". Function contracts are mostly always written as `TypesF::Arrow(.., ..)`, which is taken care of by the first case.

Here, it seems we should always return `CompletionItemKind::Property`.